src/jdk.compiler/share/classes/com/sun/tools/javac/comp/InferenceContext.java

```diff
@@ -258,7 +276,23 @@ List<Type> asInstTypes(List<Type> ts) {
      * Add custom hook for performing post-inference action
      */
     void addFreeTypeListener(List<Type> types, FreeTypeListener ftl) {
-        freeTypeListeners.put(ftl, freeVarsIn(types));
+        List<Type> myFreeVars = freeVarsIn(types);
+        if (parentIC == null) {
```

Comment:
I'm not sure I get this. In the sense: I get what the code is doing, but I'm not sure I understand why it's needed. E.g. the problem we're trying to solve is when an outer method has some inference variables that have not been propagated. This means that when the outer inference context is solved, calling `asInstTypes` on that outer context won't replace all the types. But your new patch to `asInstTypes` collects (correctly) all the inference variables from all parent contexts -- and uses those to do the type subst. Isn't that enough?

I'm a bit surprised that we also need to be picky when adding listeners -- as I'd assume that once the outer inference context is solved, all the nested inference contexts will also be solved too?